### DISCUSSION TITLE
Airgap 2.3.180 tcpreplay or so-tcpreplay

### DISCUSSION THREAD

**Discussion by virginiait** [Security-Onion-Solutions/securityonion#9108] [Unsupported Versions] [answered]
My questions:
1) In an Airgapped distributed setup, either on the Manager or the Heavy nodes, is there a way to run (tcpreplay or so-tcpreplay (to replay PCAP as "new" traffic")) without needing Internet access to download the Docker Image mentioned in the link below?
2) Alternatively, am I able to download the Docker Image as a single file and sneakernet it over to the Airgapped setup.

* Replaying pcap trace-files against the Manager or Heavy node's interfaces (from another physical computer) is not an option.
** Replay must be from/on the box itself using the sample tracefiles placed on the box.

Setup
Security Onion Version: 2.3.180 (ISO'd)
Installation: Distributed Architecture on-prem w/no internet access/connection (AirGapped). Point-to-Point connected. All nodes up'n green in GRID (except the AnalystVMs of course). TAPs only. No SPANs.
	- Manager (Manager/Search) Node (1)
	- Heavy (Sensor/Search) Node (1+)
	- AnalystVMs (1+)
Hardware (Manager & Search Nodes): Racked Servers w/ full stack of RAM; maxed HDD.
Hardware (AnalystVMs): Laptops having VMWare w/ample RAM/HDD.
No issues monitoring traffic or pivoting between HW or w/in the SOC. No failures. :-)

I've saw: https://docs.securityonion.net/en/2.3/pcaps.html?highlight=pcap
I've downloaded Wireshark trace files from: https://wiki.wireshark.org/SampleCaptures & sneakernet'd them into the Airgapped setup.
I attempted to use tcpreplay & so-tcpreplay to replay the trace files. Neither worked.
- Prompted with: "Reply functionality not enabled: attempting to enable now (may require Internet access).. Pulling so-tcpreplay image". (Times out and returns back to the terminal)
- Looking at the log file: /opt/so/log/tcpreplay/init.log I see
	"curl: (6) Could not resolve host: raw.githubusercontent.com; Unknown error
	Command failed with error 6
	Could not pull signature key file, please ensure connectivity to https://raw.githubusercontent.com"
- However, so-import-pcap did work (somewhat). Although, as stated in SO docs, it kept the original timestamps.
- I need to run/rerun trace files with present day/time values (e.g. replay the pcap as new traffic)

I saw: https://docs.securityonion.net/en/2.3/so-test.html?highlight=so-tcpreplay
And noticed: "...... if you have a distributed deployment, make sure you run so-tcpreplay on the manager first to download the necessary Docker image."
- As the setup is Airgapped. I'm not able to download the Docker image.

Lastly, thanks for making this awesome tool available and your tireless contributions to the community.

**Comment by dougburks** [comment]
You could sneakernet the so-tcpreplay image or you might just be able to find a CentOS 7 rpm for tcpreplay.

**Comment by virginiait** [comment]
Hey Doug,

Thanks for the suggestions. I chose the option to spin up another SecOnion ISO instance w/a monitored internet connection and launched the following to pull down the Docker image:
#sudo so-test
It eventually pulled the Docker image so-tcpreplay:2.3.180 from the web, installed locally, and auto replayed test PCAPs. So, that went well.

For sneaker-netting over to our air-gapped environment, I saved a copy of the Docker on that same SecOnion ISO'd instance by running:
#sudo docker save -o tcpreplay.tar sensorname:5000/security-onion-solutions/so-tcpreplay:2.3.180

I walked the saved copy with a miscellaneous name of (tcpreplay.tar) to the air-gapped system, then ran the following to load up (tcpreplay.tar):
#sudo docker load -i tcpreplay.tar

To verify it loaded on the air-gapped system, I then ran the following:
#sudo docker images

It displayed (amongst the other dockers):
--------------
REPOSITORY						                                 SIZE	TAG	        IMAGE ID 		CREATED
sensorname:5000/security-onion-solutions/so-tcpreplay	998MB	2.3.180	6f9ae~alpha_nums	6 weeks ago

So, that went well as well. Big thanks!  :-)

However, during the transfer, AntiVirus hit on a few items within the tcpreplay.tar. Specifically under... tcpreplay.tar\layer.tar\*.pcaps and the *.js's w/in a few of the PCAPs.

Obviously some of these are PCAP samples used for testing IDS sigs (eg: markfu and mta) w/in the Docker image, as I can see where they ultimately reside on each system once the Docker is loaded, their locations (not all) for example are under:
/var/lib/docker/overlay2/7ce8c8~alpha_nums/diff/opt/samples/markofu/~
and
/var/lib/docker/overlay2/7ce8c8~alpha_nums/diff/opt/samples/mta/~

- Curious though, did I run the incorrect command at the beginning, or would there be a cleaner copy of the so-tcpreplay:2.3.180 Docker image (that does not include these types of pcaps (or any pcaps))? Our IA folks get super excited in air-gapped environments when things pop off, even when we know the source. ;-)

- If the option does not currently exist, may I request a so-tcpreplay:2.3180+ docker image to be made available - with no PCAPS (or maybe ones that contain just network trace files as an alternative, that will not flag AV. https://wiki.wireshark.org/SampleCaptures).

Thank you.

**Comment by dougburks** [comment, marked as answer]
If you just want tcpreplay without pcap samples, then you're probably better off with the other option I mentioned above which was sneakernetting just the `tcpreplay` RPM package.

**Comment by virginiait** [comment]
Downloaded the rpm and brought it over worked. For others:

Filename: 	tcpreplay-4.4.2-1.el7.x86_64.rpm (and downloaded to sneaker net media)
From: 	 	https://download-ib01.fedoraproject.org/pub/epel/7/x86_64/Packages/t/tcpreplay-4.4.2-1.el7.x86_64.rpm
Installed onto standalone via:   # sudo rpm -ivh tcpreplay-4.4.2-1.el7.x86_64.rpm
Tested:		# sudo tcpreplay -i <interface_name> /path_to_WireShark_trace_files_on_local_computer/*.pcap
			Wireshark trace files previously copied from https://wiki.wireshark.org/SampleCaptures
Result:		# Replayed WireShark trace files with recent timestamps

Thanks
